### PR TITLE
Firestore backend

### PR DESCRIPTION
**Pull request by adamallred**
This PR:

* Adds Google Firestore as a backend
  * Add `firestore` as a `backend` command line switch, with leveldb as default, providing backward compatibility
* Updates the dump loader:
  * To use structs with tags that are compatible with the current dump format provided at `api/urls`
  * To specify HTTP or HTTPS protocol for the destination endpoint
  * To close upload response bodies after POSTing to avoid exhausting file handles for large dump uploads

### CODE REVIEW

File: backend/firestore/firestore.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+type Backend struct {
+	db *fs.Client
+}
+
+// New instantiates a new Backend
+func New(path string) (*Backend, error) {
+	if path == "" {
+		path = getGoogleProject()
+	}
+	ctx := context.Background()
```
Comment (kellegous) [inline backend/firestore/firestore.go]: `ctx` should probably be passed in here, right?

File: backend/firestore/firestore.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+// Backend provides access to Google Firestore.
+type Backend struct {
+	db *fs.Client
+}
+
+// New instantiates a new Backend
+func New(path string) (*Backend, error) {
+	if path == "" {
+		path = getGoogleProject()
```
Comment (kellegous) [inline backend/firestore/firestore.go]: If you pass in the `Context`, you can also propagate it to this.

File: backend/firestore/firestore.go (hunk 1 above)
Comment (adamallred) [reply backend/firestore/firestore.go]: It seems like `context.Background()` is the only reasonable value for the `New` call. Setting any timeout or deadline causes use of the client after that to fail with deadline exceeded. I guess I can just build that in `main.go`.

File: backend/firestore/firestore.go (hunk 2 above)
Comment (adamallred) [reply backend/firestore/firestore.go]: The passed in context will be `context.Background()` with no timeout; I wasn't wanting that to be the case for this call.

**Pull request merged**